src/ratelimit.rs

```diff
@@ -0,0 +1,137 @@
+//! # Rate limiting module.
+//!
+//! This module contains implementation of token bucket policy.
+//! Its primary use is preventing Delta Chat from sending too many messages, especially automatic,
+//! such as read receipts.
+
+use std::time::{Duration, SystemTime};
+
+#[derive(Debug)]
+pub(crate) struct Ratelimit {
+    /// Time of the last update.
+    last_update: SystemTime,
+
+    /// Number of messages sent within the time window ending at `last_update`.
+    current_value: f64,
+
+    /// Time window size.
+    window: Duration,
+
+    /// Number of messages allowed to send within the time window.
+    quota: f64,
+}
+
+impl Ratelimit {
+    /// Returns a new rate limiter with the given constraints.
+    ///
+    /// Rate limiter will allow to send no more than `quota` messages within duration `window`.
+    pub(crate) fn new(window: Duration, quota: f64) -> Self {
+        Self::new_at(window, quota, SystemTime::now())
+    }
+
+    /// Returns a new rate limiter with given current time for testing purposes.
+    const fn new_at(window: Duration, quota: f64, now: SystemTime) -> Self {
+        Self {
+            last_update: now,
+            current_value: 0.0,
+            window,
+            quota,
+        }
+    }
+
+    /// Returns current number of sent messages.
+    fn current_value_at(&self, now: SystemTime) -> f64 {
+        let rate: f64 = self.quota / self.window.as_secs_f64();
+        let elapsed = now
+            .duration_since(self.last_update)
+            .unwrap_or(Duration::ZERO)
+            .as_secs_f64()
+            .max(0.0);
+        (self.current_value - rate * elapsed).max(0.0)
```

Comment:
Nit (feel free to ignore): `x.max(y)` doesn't do what at least I would think: it computes the maximum of x and y (i.e. `max(x, y)`), while I would expect that it limits x to a maximum value of y (which would be `min(x, y)`).

Since a Duration can't be negative anyway, what about:
```suggestion
        f64::max(self.current_value - rate * elapsed, 0.0)
```

Reply:
`elapsed` can be negative, because it is `SystemTime` and it's not guaranteed to be monotonic.